plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+-- the other generators.
+runGenTmCustom :: Int
+               -> (Maybe (Type TyName DefaultUni ())
+                  -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ()))
+               -> GenTm a
+               -> Gen a
+runGenTmCustom f cg g = do
+  sized $ \ n -> do
+    let env = GenEnv
+          { geSize               = n
+          , geDatas              = Map.empty
+          , geTypes              = Map.empty
+          , geTerms              = Map.empty
+          , geUnboundUsedTyNames = Set.empty
+          , geEscaping           = YesEscape
+          , geCustomGen          = cg
+          , geCustomFreq         = f
+          , geDebug              = False
+          }
+    flip runReader env <$> runGenT g
+
+suchThatMap :: Monad m => GenT m a -> (a -> Maybe b) -> GenT m b
+gen `suchThatMap` f = fmap fromJust $ fmap f gen `suchThat` isJust
+
+deliver :: Monad m => GenT m (Maybe a) -> GenT m a
+deliver gen = gen `suchThatMap` id
+
+-- * Utility functions
+
+-- | Don't allow types to escape from a generator.
+withNoEscape :: GenTm a -> GenTm a
+withNoEscape = local $ \env -> env { geEscaping = NoEscape }
+
+-- * Dealing with size
+
+-- | Map a function over the generator size
+onSize :: (Int -> Int) -> GenTm a -> GenTm a
+onSize f = local $ \ env -> env { geSize = f (geSize env) }
+
+-- | Default to the first generator if the size is zero (or negative),
+-- use the second generator otherwise.
+ifSizeZero :: GenTm a -> GenTm a -> GenTm a
+ifSizeZero ifZ nonZ = do
+  n <- asks geSize
+  if n <= 0 then ifZ else nonZ
+
+-- | Locally set the size in a generator
+withSize :: Int -> GenTm a -> GenTm a
+withSize = onSize . const
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments.
+sizeSplit_ :: Int -> Int -> GenTm a -> GenTm b -> GenTm (a, b)
+sizeSplit_ a b ga gb = sizeSplit a b ga (const gb)
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments and use the result of the first generator
+-- in the second.
+sizeSplit :: Int -> Int -> GenTm a -> (a -> GenTm b) -> GenTm (a, b)
+sizeSplit a b ga gb = do
+  n <- asks geSize
+  let na = (a * n) `div` (a + b)
+      nb = (b * n) `div` (a + b)
+  x <- withSize na ga
+  (,) x <$> withSize nb (gb x)
+
+-- * Names
+
+-- | Get all uniques we have generated and are used in the current context.
+getUniques :: GenTm (Set Unique)
+getUniques = do
+  GenEnv{geDatas = dts, geTypes = tys, geTerms = tms, geUnboundUsedTyNames = used} <- ask
+  return $ Set.mapMonotonic (_nameUnique . unTyName) (Map.keysSet dts <> Map.keysSet tys <> used) <>
+           Set.mapMonotonic _nameUnique (Map.keysSet tms) <>
+           Set.unions [ names d | d <- Map.elems dts ]
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
+
+{- Note [Warning about generating fresh names]: because `GenTm` is a *reader* monad
```

Comment:
I think the Note is a bit confusing, I'll polish it while I'm here. We can't avoid what the Note describes, because we have to model scopes and for that we need a reader monad and for that we have to have some explicit for of binding variables. The mentioned in the Note overlap thing is different and for that we could indeed use `Quote`, but see PLT-806, at this point it's quite hard to change the logic and I definitely don't want this to be a blocker, giving that the current behavior is not wrong, just not optimal.